server/src/main/java/org/elasticsearch/action/support/FanOutListener.java

```diff
@@ -0,0 +1,294 @@
+                continue;
+            }
+            if (currentValue instanceof Cell head) {
+                if (newCell == null) {
+                    newCell = new Cell(wrappedListener, head);
+                } else {
+                    newCell.next = head;
+                }
+                currentValue = ref.compareAndExchange(head, newCell);
+                if (currentValue == head) {
+                    return;
+                }
+            } else {
+                assert false : "unexpected witness: " + currentValue;
+            }
+        }
+    }
+
+    @Override
+    public final void onResponse(T result) {
+        setResult(new SuccessResult<T>(result));
+    }
+
+    @Override
+    public final void onFailure(Exception exception) {
+        setResult(new FailureResult(exception, wrapException(exception)));
+    }
+
+    protected Exception wrapException(Exception exception) {
+        return exception;
+    }
+
+    /**
+     * @return {@code true} if and only if this listener has been completed (either successfully or exceptionally).
+     */
+    public final boolean isDone() {
+        return isDone(ref.get());
+    }
+
+    /**
+     * @return the result with which this listener completed successfully, or throw the exception with which it failed.
+     *
+     * @throws AssertionError if this listener is not complete yet and assertions are enabled.
+     * @throws IllegalStateException if this listener is not complete yet and assertions are disabled.
+     */
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    public final T rawResult() throws Exception {
+        final var refValue = ref.get();
+        if (refValue instanceof SuccessResult result) {
+            return (T) result.result();
+        } else if (refValue instanceof FailureResult result) {
+            throw result.exception();
+        } else {
+            assert false : "not done";
+            throw new IllegalStateException("listener is not done, cannot get result yet");
+        }
+    }
+
+    protected static RuntimeException wrapAsExecutionException(Throwable t) {
+        if (t instanceof RuntimeException runtimeException) {
+            return runtimeException;
+        } else {
+            return new UncategorizedExecutionException("Failed execution", new ExecutionException(t));
+        }
+    }
+
+    private static <T> ActionListener<T> preserveContext(@Nullable ThreadContext threadContext, ActionListener<T> listener) {
+        return threadContext == null ? listener : ContextPreservingActionListener.wrapPreservingContext(listener, threadContext);
+    }
+
+    private static <T> ActionListener<T> fork(Executor executor, ActionListener<T> listener) {
+        return executor == EsExecutors.DIRECT_EXECUTOR_SERVICE ? listener : new ThreadedActionListener<>(executor, listener);
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    private static <T> boolean tryComplete(Object refValue, ActionListener<T> listener) {
```

Comment:
Presumably the executing thread doesn't matter here? It could be the `addListener` thread, or the dependent task's executing thread, depending on when exactly the listener was added. Does this change the thread behaviour at all?

Reply:
The executing thread does matter in some situations, but again this is how `ListenableFuture` behaves today.